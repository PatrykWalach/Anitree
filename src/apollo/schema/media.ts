import { ScoreFormat } from './viewer'
import { MediaList } from './mediaListCollection'

export interface Variables {
  isAdult?: boolean
  id: number
  format?: ScoreFormat
}

export interface Node {
  id: number
  type: MediaType
}

export type MediaType = 'MANGA' | 'ANIME'
export interface FuzzyDate {
  year: number | null
  month: number | null
  day: number | null
}

export interface Media extends Node {
  startDate: FuzzyDate
  endDate: FuzzyDate
  format: string | null
  description: string | null
  season: string | null
  seasonInt: number | null
  episodes: number | null
  chapters: number | null
  volumes: number | null
  isFavourite: boolean
  mediaListEntry: MediaList | null
  title: MediaTitle
  siteUrl: string
  relations: {
    edges: MediaEdge[]
  }
  tags: {
    rank: number
    name: string
  }[]
  studios: {
    nodes: Studio[]
  }
  status: MediaStatus | null
  bannerImage: string | null
  coverImage: MediaCoverImage
  stats: MediaStats
}
export interface MediaStats {
  scoreDistribution: ScoreDistribution[] | null
}
export interface ScoreDistribution {
  score: number
  amount: number
}
export interface MediaTitle {
  romaji: string
  english: string
  native: string
}

export type MediaRelation =
  | 'ADAPTATION'
  | 'PREQUEL'
  | 'SEQUEL'
  | 'PARENT'
  | 'SIDE_STORY'
  | 'CHARACTER'
  | 'SUMMARY'
  | 'ALTERNATIVE'
  | 'SPIN_OFF'
  | 'OTHER'
  | 'SOURCE'
  | 'COMPILATION'
  | 'CONTAINS'

export interface MediaEdge {
  node: Node
  relationType: MediaRelation
  id: number
}

export interface Studio {
  id: number
  name: string
}

export interface MediaCoverImage {
  extraLarge: string | null
  large: string | null
  medium: string | null
  color: string | null
}

export type MediaStatus =
  | 'CURRENT'
  | 'PLANNING'
  | 'COMPLETED'
  | 'DROPPED'
  | 'PAUSED'
  | 'REPEATING'
