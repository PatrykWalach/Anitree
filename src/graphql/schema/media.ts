import { ScoreFormat } from './viewer'
import { MediaList } from './mediaListCollection'

export interface Variables {
  isAdult?: boolean
  id: number
  format?: ScoreFormat
}

export interface Node {
  __typename: 'Media'
  id: number
  type: MediaType
}

export type MediaType = 'MANGA' | 'ANIME'
export interface FuzzyDate {
  __typename: 'FuzzyDate'
  year: number | null
  month: number | null
  day: number | null
}

export interface Media extends Node {
  startDate: FuzzyDate
  endDate: FuzzyDate
  format: string | null
  description: string | null
  season: string | null
  seasonInt: number | null
  episodes: number | null
  chapters: number | null
  volumes: number | null
  isFavourite: boolean
  ranknigs: MediaRank[]
  mediaListEntry: MediaList | null
  title: MediaTitle
  siteUrl: string
  relations: MediaConnection
  tags: MediaTag[]
  studios: StudioConnection
  status: MediaStatus | null
  bannerImage: string | null
  coverImage: MediaCoverImage
  stats: MediaStats
}
export interface MediaConnection {
  __typename: 'MediaConnection'
  edges: MediaEdge[]
}
export interface MediaTag {
  __typename: 'MediaTag'
  id: number
  rank: number
  name: string
}
export interface StudioConnection {
  __typename: 'StudioConnection'

  nodes: Studio[]
}
export interface MediaRank {
  __typename: 'MediaRank'
  allTime: boolean
  context: string
  rank: number
  type: 'RATED' | 'POPULAR'
  year: null | number
}
export interface MediaStats {
  __typename: 'MediaStats'
  scoreDistribution: ScoreDistribution[] | null
}
export interface ScoreDistribution {
  __typename: 'ScoreDistribution'
  score: number
  amount: number
}
export interface MediaTitle {
  __typename: 'MediaTitle'
  romaji: string
  english: string
  native: string
}

export type MediaRelation =
  | 'ADAPTATION'
  | 'PREQUEL'
  | 'SEQUEL'
  | 'PARENT'
  | 'SIDE_STORY'
  | 'CHARACTER'
  | 'SUMMARY'
  | 'ALTERNATIVE'
  | 'SPIN_OFF'
  | 'OTHER'
  | 'SOURCE'
  | 'COMPILATION'
  | 'CONTAINS'

export interface MediaEdge {
  __typename: 'MediaEdge'
  node: Node
  relationType: MediaRelation
  id: number
}

export interface Studio {
  __typename: 'Studio'
  id: number
  name: string
}

export interface MediaCoverImage {
  __typename: 'MediaCoverImage'
  extraLarge: string | null
  large: string | null
  medium: string | null
  color: string | null
}

export type MediaStatus =
  | 'CURRENT'
  | 'PLANNING'
  | 'COMPLETED'
  | 'DROPPED'
  | 'PAUSED'
  | 'REPEATING'
