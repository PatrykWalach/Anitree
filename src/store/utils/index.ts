export * from './useCommands'
export * from './useTitle'
export * from './useEdit'
