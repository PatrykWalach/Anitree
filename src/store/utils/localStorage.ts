export const getItem = <T>(key: string, defaultValue?: T) => {
  const cache = localStorage.getItem(key)
  const stored = cache && JSON.parse(cache)

  return {
    ...defaultValue,
    ...stored,
  } as T
}
export const setItem = <T>(key: string, value: T) =>
  localStorage.setItem(key, JSON.stringify({ ...value }))
