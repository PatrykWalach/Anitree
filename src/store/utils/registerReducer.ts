// import * as reducers from '../reducers'
// import { Reducer, ReducersMapObject, combineReducers } from 'redux'
// import { EnhancedStore } from '@reduxjs/toolkit'
// import { store, useStore } from '../'

// const staticReducers: ReducersMapObject<any, any> = reducers
// const asyncReducers: ReducersMapObject<any, any> = {}

// export const registerReducer = <K extends string, R extends Reducer<any>>(
//   key: K,
//   asyncReducer: R,
// ) => {
//   asyncReducers[key] = asyncReducer
//   store.replaceReducer(createReducer(asyncReducers) as any)

//   return () =>
//     useStore<
//       typeof store extends EnhancedStore<infer S>
//         ? EnhancedStore<S & { [key in K]: R }>
//         : any
//     >()
// }

// export const unregisterReducer = <K extends string>(key: K) => {
//   delete asyncReducers[key]
//   store.replaceReducer(createReducer(asyncReducers) as any)
//   return () =>
//     useStore<
//       typeof store extends EnhancedStore<infer S>
//         ? EnhancedStore<Omit<S, K>>
//         : any
//     >()
// }

// const createReducer = (asyncReducers: ReducersMapObject<any, any> = {}) =>
//   combineReducers({
//     ...staticReducers,
//     ...asyncReducers,
//   })
