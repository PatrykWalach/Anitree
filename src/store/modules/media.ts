import Vue from 'vue'
import { apolloQueryPage } from '../api'
import {
  Module,
  VuexModule,
  Mutation,
  Action,
  getModule,
  MutationAction
} from 'vuex-module-decorators'
import store from '@/store'
import {
  MediaNode,
  Media,
  MediaEdgeExtended,
  MediaEdge,
  Data,
  RawFilter,
  Filter,
  FetchVariables,
  Page
} from '@/types'

function createVariables(
  acc: {
    idIn: number[]
  }[],
  media: MediaNode
): {
  idIn: number[]
}[] {
  const index = acc.findIndex(arr => arr.idIn.length < 50)
  if (index === -1) {
    acc.push({ idIn: [media.id] })
  } else {
    acc[index].idIn.push(media.id)
  }
  return acc
}

function applyFilter(
  filteredEdges: MediaEdgeExtended[],
  filter: RawFilter
): MediaEdgeExtended[] {
  return filteredEdges.filter(filter)
}

function filterDuplicate<E = any>(
  test: (element: E, otherElement: E, i?: number, arr?: E[]) => boolean,
  element: E,
  i: number,
  arr: E[]
): boolean {
  return arr.findIndex(test.bind(null, element)) === i
}

function nodesAreEqual(element: MediaNode, otherElement: MediaNode) {
  return element.id === otherElement.id
}

function filterDuplicateNodes(
  node: MediaNode,
  i: number,
  allNodes: MediaNode[]
): boolean {
  return filterDuplicate(nodesAreEqual, node, i, allNodes)
}

function filterDuplicateEdges(
  edge: MediaEdge,
  i: number,
  allEdges: MediaEdge[]
): boolean {
  return filterDuplicateNodes(edge.node, i, allEdges.map(({ node }) => node))
}
@Module({
  namespaced: true,
  name: 'media',
  store,
  dynamic: true
})
export class ModuleMedia extends VuexModule {
  public currentId: number | null = null

  public media: Data<Media> = {}

  public filters: Data<Filter> = {}

  // public filteredMedia: MediaEdgeExtended[] = []

  public get activeFilters() {
    return Object.values(this.filters).filter(({ active }) => active)
  }

  public get exclusiveFilters(): RawFilter[] {
    return this.activeFilters
      .filter(({ exclusive }) => exclusive)
      .map(({ filter }) => filter)
  }

  public get inclusiveFilters(): RawFilter[] {
    return this.activeFilters
      .filter(({ exclusive }) => !exclusive)
      .map(({ filter }) => filter)
  }

  public get currentMedia(): Media | null {
    const { currentId, media } = this
    return (currentId && media[currentId]) || null
  }

  public get relationTypes(): string[] {
    const { inclusiveFilters, filteredMedia } = this
    return [
      ...new Set(
        inclusiveFilters
          .reduce(applyFilter, filteredMedia)
          .flatMap(media => media.node.relations.edges)
          .map(({ relationType }) => relationType)
      )
    ]
  }

  public get sortedMedia(): Media[] {
    const getYear = (seasonInt: number) => {
      const year = Math.floor(seasonInt / 10)
      return year > 50 ? 1900 + year : 2000 + year
    }

    return this.filteredMedia
      .map(({ node }) => node)
      .sort(
        (
          { startDate: dateA, seasonInt: seasonA },
          { startDate: dateB, seasonInt: seasonB }
        ) =>
          (seasonA &&
            seasonB &&
            (getYear(seasonA) - getYear(seasonB) ||
              (seasonA % 10) - (seasonB % 10))) ||
          (dateA.year && dateB.year && dateA.year - dateB.year) ||
          (dateA.month && dateB.month && dateA.month - dateB.month) ||
          (dateA.day && dateB.day && dateA.day - dateB.day) ||
          1
      )
  }

  public get filteredMedia(): MediaEdgeExtended[] {
    const getRelatedMedia = (
      edges: MediaEdge[],
      acc: MediaEdgeExtended[] = []
    ): MediaEdgeExtended[] => {
      const { exclusiveFilters, inclusiveFilters, media } = this

      const newMedia = exclusiveFilters.reduce(
        applyFilter,
        edges
          .map(({ relationType, node }) => {
            return {
              relationType,
              node: media[node.id]
            }
          })
          .filter(({ node }) => node)
      )

      acc.push(...newMedia)
      const accNodes = acc.map(({ node }) => node)
      const relatedNodes = inclusiveFilters
        .reduce(applyFilter, newMedia)
        .map(({ node }) => node)
        .flatMap(({ relations }) => relations.edges)
        .filter(filterDuplicateEdges)
        .filter(({ node }) => !accNodes.some(nodesAreEqual.bind(null, node)))

      if (relatedNodes.length) {
        return getRelatedMedia(relatedNodes, acc)
      }

      return acc
    }

    const { currentId } = this

    return currentId
      ? getRelatedMedia([
          {
            node: {
              id: currentId,
              type: ''
            },
            relationType: '',
            id: -1
          }
        ])
      : []
  }

  @Mutation
  public CHANGE_FILTER(filter: Filter): Filter {
    return Vue.set(this.filters, filter.id, filter)
  }

  @Mutation
  public ADD_MEDIA(media: Media): Media {
    return Vue.set(this.media, media.id, media)
  }

  @MutationAction
  public async CHANGE_CURRENT_ID({
    currentId
  }: {
    currentId: number | null
  }): Promise<{
    currentId: number | null
  }> {
    return { currentId }
  }

  @Action
  public async extendMediaEdge({
    relationType,
    node
  }: MediaEdge): Promise<MediaEdgeExtended> {
    return Promise.resolve({
      relationType,
      node: await this.getMedia(node.id)
    })
  }

  @Action
  public async fetchMedia(variables: FetchVariables): Promise<Media[]> {
    return this.fetchMediaPage(variables).then(page => page.media)
  }

  @Action
  public async fetchMediaPage(variables: FetchVariables): Promise<Page> {
    return new Promise(resolve =>
      apolloQueryPage(variables)
        .then(page => {
          const newMedia = page.media
          newMedia.forEach(this.ADD_MEDIA)
          return page
        })
        .then(resolve)
        .catch(() =>
          setTimeout(() => this.fetchMediaPage(variables).then(resolve), 10000)
        )
    )
  }

  @Action
  public async handleQueue(medias: Media[]): Promise<void> {
    return new Promise(resolve => {
      const { media } = this
      const newMedias: MediaNode[] = medias
        .flatMap(newMedia => newMedia.relations.edges)
        .map(({ node }) => node)
        .filter(filterDuplicateNodes)
        .filter(newMedia => !media[newMedia.id])

      if (newMedias.length) {
        const { fetchMedia, handleQueue } = this
        Promise.all(newMedias.reduce(createVariables, []).map(fetchMedia))
          .then(result => result.flat())
          .then(handleQueue)
          .then(resolve)
      } else {
        resolve()
      }
    })
  }

  // @Action
  // public async getFilteredMedia({
  //   edges = [
  //     {
  //       node: {
  //         id: this.currentId || -1,
  //         type: ''
  //       },
  //       relationType: '',
  //       id: -1
  //     }
  //   ],
  //   acc = []
  // }: {
  //   edges?: MediaEdge[]
  //   acc?: MediaEdgeExtended[]
  // }): Promise<MediaEdgeExtended[]> {
  //   return new Promise(async resolve => {
  //     const {
  //       exclusiveFilters,
  //       inclusiveFilters
  //       // ,media
  //     } = this

  //     const newMedia = exclusiveFilters.reduce(
  //       applyFilter,
  //       await Promise.all(edges.map(this.extendMediaEdge))
  //       // edges
  //       //   .map(({ relationType, node }) => {
  //       //     return {
  //       //       relationType,
  //       //       node: media[node.id]
  //       //     }
  //       //   })
  //       //   .filter(({ node }) => node)
  //     )

  //     acc.push(...newMedia)
  //     const accNodes = acc.map(({ node }) => node)
  //     const relatedNodes = inclusiveFilters
  //       .reduce(applyFilter, newMedia)
  //       .map(({ node }) => node)
  //       .flatMap(({ relations }) => relations.edges)
  //       .filter(filterDuplicateEdges)
  //       .filter(({ node }) => !accNodes.some(nodesAreEqual.bind(null, node)))

  //     if (relatedNodes.length) {
  //       this.getFilteredMedia({ edges: relatedNodes, acc }).then(resolve)
  //     } else {
  //       resolve(acc)
  //     }
  //   })
  // }

  // @Mutation
  // public CHANGE_FILTERED_MEDIA(filteredMedia: MediaEdgeExtended[]) {
  //   this.filteredMedia = filteredMedia
  // }

  // @Action({ commit: 'CHANGE_FILTERED_MEDIA' })
  // public async changeFilteredMedia() {
  //   return await this.getFilteredMedia({})
  // }

  @Action
  public async getMedia(id: number) {
    return new Promise<Media>(async (resolve, reject) => {
      const storedMedia = this.media[id]
      if (storedMedia) {
        resolve(storedMedia)
      } else {
        const media = await this.fetchMedia({
          idIn: [id]
        })

        if (media.length) {
          resolve(media.shift())
        } else {
          reject()
        }
      }
    })
  }
}
export const media =getModule(ModuleMedia)
export default media
