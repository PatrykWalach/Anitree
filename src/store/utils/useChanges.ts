import {
  DELETE_MEDIA_LIST_ENTRY,
  MEDIA,
  SAVE_MEDIA_LIST_ENTRY,
} from '@/graphql'
import { DeletePending, SavePending } from '../modules/changes'
import {
  DeleteVariables,
  Deleted,
  SaveVariables,
} from '@/graphql/schema/listEntry'
import { DocumentNode, FetchResult } from 'apollo-link'
import { Ref, ref } from '@vue/composition-api'
import { DataProxy } from 'apollo-cache'
import { Media } from '@/graphql/schema/media'
import { MediaList } from '@/graphql/schema/mediaListCollection'
import { MutationUpdaterFn } from 'apollo-client'
import { Variables } from '@/graphql/schema/media'
import { useActions } from '..'
import { useDispatch } from 'vue-redux-hooks'
import { useMutation } from '@vue/apollo-composable'

type OnError = (
  fn: (param?: Error | undefined) => void,
) => {
  off: () => void
}

export const handleError = <
  P extends Omit<SavePending, 'id'> | Omit<DeletePending, 'id'>
>(
  onError: OnError,
  type: P['type'],
) => {
  const pending: Ref<P> = ref({ type })

  const dispatch = useDispatch()
  const actions = useActions()

  onError(() => {
    dispatch(actions.changes.UNSHIFT_PENDING(pending.value))
  })

  return pending
}

export const useChange = <D, V>(
  query: DocumentNode,
  update: (cache: DataProxy, { data }: FetchResult<D>, mediaId: number) => void,
) => {
  const { mutate: change, loading, onError, onDone } = useMutation<D, V>(query)
  const mutate = (variables: V, mediaId: number) => {
    change(variables, {
      update: (cache, result) => update(cache, result, mediaId),
    })
  }
  return { loading, mutate, onDone, onError }
}

export const useDelete = () => {
  return useChange<{ DeleteMediaListEntry: Deleted }, DeleteVariables>(
    DELETE_MEDIA_LIST_ENTRY,
    updateDeleteMediaListEntry,
  )
}

export const useSave = () => {
  return useChange<
    {
      SaveMediaListEntry: MediaList
    },
    SaveVariables
  >(SAVE_MEDIA_LIST_ENTRY, updateSaveMediaListEntry)
}

export const useDeleteWithChanges = () => {
  const query = useDelete()

  const pending = handleError(query.onError, 'DELETE')

  const mutate = (variables: DeleteVariables, mediaId: number) => {
    pending.value = {
      mediaId,
      type: 'DELETE',
      variables,
    }

    query.mutate(variables, mediaId)
  }

  return { ...query, mutate }
}
export const useSaveWithChanges = () => {
  const query = useSave()

  const pending = handleError(query.onError, 'SAVE')

  const mutate = (variables: SaveVariables) => {
    pending.value = {
      mediaId: variables.mediaId,
      type: 'SAVE',
      variables,
    }

    query.mutate(variables, variables.mediaId)
  }

  return { ...query, mutate }
}

export const useChanges = () => {
  const { mutate: deleteEntry } = useDelete()
  const { mutate: saveEntry } = useSave()

  const isSavePending = (
    pending: SavePending | DeletePending,
  ): pending is SavePending => pending.type === 'SAVE'

  return { deleteEntry, isSavePending, saveEntry }
}

export const updateSaveMediaListEntry: MutationUpdaterFn<{
  SaveMediaListEntry: MediaList
}> = (cache, { data: result }) => {
  if (!result) return

  const query: {
    variables: Variables
    query: DocumentNode
  } = {
    query: MEDIA,
    variables: { id: result.SaveMediaListEntry.mediaId },
  }

  const data: { Media: Media } | null = cache.readQuery(query)

  if (data) {
    data.Media.mediaListEntry = result.SaveMediaListEntry

    cache.writeQuery({
      ...query,
      data,
    })
  }
}

const updateDeleteMediaListEntry = (
  cache: DataProxy,
  { data: result }: FetchResult<{ DeleteMediaListEntry: Deleted }>,
  id: number,
) => {
  if (!result || !result.DeleteMediaListEntry.deleted) return

  const query: {
    variables: Variables
    query: DocumentNode
  } = {
    query: MEDIA,
    variables: {
      id,
    },
  }

  const data: { Media: Media } | null = cache.readQuery(query)

  if (data) {
    data.Media.mediaListEntry = null

    cache.writeQuery({
      ...query,
      data,
    })
  }
}
