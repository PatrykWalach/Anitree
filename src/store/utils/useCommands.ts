import { ListCommand } from '../modules/commands'
import { computed } from '@vue/composition-api'
import { useState } from './useState'
import { useStore } from './useStore'

export const useCommands = () => {
  const { dispatch, actions } = useStore()

  const history = useState(state => state.commands.history)

  const undo = async () => {
    history.value[history.value.length - 1]
      .undo()
      .then(() => dispatch(actions.commands.POP_COMMAND()))
  }

  const add = async (command: ListCommand) => {
    dispatch(actions.commands.PUSH_COMMAND(command))
    await command.execute()
    return command
  }

  const getPending = computed(() => history.value.filter(({ done }) => !done))
  return { add, getPending, undo }
}
