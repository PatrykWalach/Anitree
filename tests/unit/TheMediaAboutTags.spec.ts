import { createVariations, matchSnapshot } from './utils'
import TheMediaAboutTags from '@/components/TheMediaAboutTags.vue'
import { mockedMedia } from './mocks/media'

describe('TheMediaAboutTags', () => {
  createVariations({
    propsData: {
      media: () => [mockedMedia, null],
    },
  }).forEach(settings => matchSnapshot(TheMediaAboutTags, settings))
})
