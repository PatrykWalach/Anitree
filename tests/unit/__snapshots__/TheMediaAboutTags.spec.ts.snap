// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheMediaAboutTags matches snapshot 1`] = `
<v-card-text-stub>
  <v-chip-stub
    active="true"
    activeclass=""
    class="text-uppercase"
    closeicon="$delete"
    filtericon="$complete"
    label="true"
    link="true"
    outlined="true"
    style="margin: 4px 8px 4px 0px;"
    tag="span"
    to="[object Object]"
  >
    
    spring 1998
  
  </v-chip-stub>
   
  <v-chip-stub
    active="true"
    activeclass=""
    closeicon="$delete"
    filtericon="$complete"
    label="true"
    link="true"
    style="margin: 4px 8px 4px 0px;"
    tag="span"
    to="[object Object]"
  >
    <v-avatar-stub
      left="true"
      size="48"
    >
       94 
    </v-avatar-stub>
    
    space
  
  </v-chip-stub>
  <v-chip-stub
    active="true"
    activeclass=""
    closeicon="$delete"
    filtericon="$complete"
    label="true"
    link="true"
    style="margin: 4px 8px 4px 0px;"
    tag="span"
    to="[object Object]"
  >
    <v-avatar-stub
      left="true"
      size="48"
    >
       92 
    </v-avatar-stub>
    
    lgbtq issues
  
  </v-chip-stub>
  <v-chip-stub
    active="true"
    activeclass=""
    closeicon="$delete"
    filtericon="$complete"
    label="true"
    link="true"
    style="margin: 4px 8px 4px 0px;"
    tag="span"
    to="[object Object]"
  >
    <v-avatar-stub
      left="true"
      size="48"
    >
       85 
    </v-avatar-stub>
    
    episodic
  
  </v-chip-stub>
  <v-chip-stub
    active="true"
    activeclass=""
    closeicon="$delete"
    filtericon="$complete"
    label="true"
    link="true"
    style="margin: 4px 8px 4px 0px;"
    tag="span"
    to="[object Object]"
  >
    <v-avatar-stub
      left="true"
      size="48"
    >
       84 
    </v-avatar-stub>
    
    ensemble cast
  
  </v-chip-stub>
  <v-chip-stub
    active="true"
    activeclass=""
    closeicon="$delete"
    filtericon="$complete"
    label="true"
    link="true"
    style="margin: 4px 8px 4px 0px;"
    tag="span"
    to="[object Object]"
  >
    <v-avatar-stub
      left="true"
      size="48"
    >
       79 
    </v-avatar-stub>
    
    crime
  
  </v-chip-stub>
  <v-chip-stub
    active="true"
    activeclass=""
    closeicon="$delete"
    filtericon="$complete"
    label="true"
    link="true"
    style="margin: 4px 8px 4px 0px;"
    tag="span"
    to="[object Object]"
  >
    <v-avatar-stub
      left="true"
      size="48"
    >
       77 
    </v-avatar-stub>
    
    primarily adult cast
  
  </v-chip-stub>
  <v-chip-stub
    active="true"
    activeclass=""
    closeicon="$delete"
    filtericon="$complete"
    label="true"
    link="true"
    style="margin: 4px 8px 4px 0px;"
    tag="span"
    to="[object Object]"
  >
    <v-avatar-stub
      left="true"
      size="48"
    >
       76 
    </v-avatar-stub>
    
    noir
  
  </v-chip-stub>
  <v-chip-stub
    active="true"
    activeclass=""
    closeicon="$delete"
    filtericon="$complete"
    label="true"
    link="true"
    style="margin: 4px 8px 4px 0px;"
    tag="span"
    to="[object Object]"
  >
    <v-avatar-stub
      left="true"
      size="48"
    >
       72 
    </v-avatar-stub>
    
    guns
  
  </v-chip-stub>
  <v-chip-stub
    active="true"
    activeclass=""
    closeicon="$delete"
    filtericon="$complete"
    label="true"
    link="true"
    style="margin: 4px 8px 4px 0px;"
    tag="span"
    to="[object Object]"
  >
    <v-avatar-stub
      left="true"
      size="48"
    >
       70 
    </v-avatar-stub>
    
    drugs
  
  </v-chip-stub>
  <v-chip-stub
    active="true"
    activeclass=""
    closeicon="$delete"
    filtericon="$complete"
    label="true"
    link="true"
    style="margin: 4px 8px 4px 0px;"
    tag="span"
    to="[object Object]"
  >
    <v-avatar-stub
      left="true"
      size="48"
    >
       70 
    </v-avatar-stub>
    
    tragedy
  
  </v-chip-stub>
  <v-chip-stub
    active="true"
    activeclass=""
    closeicon="$delete"
    filtericon="$complete"
    label="true"
    link="true"
    style="margin: 4px 8px 4px 0px;"
    tag="span"
    to="[object Object]"
  >
    <v-avatar-stub
      left="true"
      size="48"
    >
       68 
    </v-avatar-stub>
    
    male protagonist
  
  </v-chip-stub>
  <v-chip-stub
    active="true"
    activeclass=""
    closeicon="$delete"
    filtericon="$complete"
    label="true"
    link="true"
    style="margin: 4px 8px 4px 0px;"
    tag="span"
    to="[object Object]"
  >
    <v-avatar-stub
      left="true"
      size="48"
    >
       60 
    </v-avatar-stub>
    
    philosophy
  
  </v-chip-stub>
  <v-chip-stub
    active="true"
    activeclass=""
    closeicon="$delete"
    filtericon="$complete"
    label="true"
    link="true"
    style="margin: 4px 8px 4px 0px;"
    tag="span"
    to="[object Object]"
  >
    <v-avatar-stub
      left="true"
      size="48"
    >
       60 
    </v-avatar-stub>
    
    martial arts
  
  </v-chip-stub>
  <v-chip-stub
    active="true"
    activeclass=""
    closeicon="$delete"
    filtericon="$complete"
    label="true"
    link="true"
    style="margin: 4px 8px 4px 0px;"
    tag="span"
    to="[object Object]"
  >
    <v-avatar-stub
      left="true"
      size="48"
    >
       58 
    </v-avatar-stub>
    
    amnesia
  
  </v-chip-stub>
  <v-chip-stub
    active="true"
    activeclass=""
    closeicon="$delete"
    filtericon="$complete"
    label="true"
    link="true"
    style="margin: 4px 8px 4px 0px;"
    tag="span"
    to="[object Object]"
  >
    <v-avatar-stub
      left="true"
      size="48"
    >
       48 
    </v-avatar-stub>
    
    primarily male cast
  
  </v-chip-stub>
  <v-chip-stub
    active="true"
    activeclass=""
    closeicon="$delete"
    filtericon="$complete"
    label="true"
    link="true"
    style="margin: 4px 8px 4px 0px;"
    tag="span"
    to="[object Object]"
  >
    <v-avatar-stub
      left="true"
      size="48"
    >
       46 
    </v-avatar-stub>
    
    anti-hero
  
  </v-chip-stub>
  <v-chip-stub
    active="true"
    activeclass=""
    closeicon="$delete"
    filtericon="$complete"
    label="true"
    link="true"
    style="margin: 4px 8px 4px 0px;"
    tag="span"
    to="[object Object]"
  >
    <v-avatar-stub
      left="true"
      size="48"
    >
       45 
    </v-avatar-stub>
    
    yakuza
  
  </v-chip-stub>
</v-card-text-stub>
`;

exports[`TheMediaAboutTags matches snapshot 2`] = `
<v-chip-group-stub
  activeclass="v-slide-item--active"
  mobilebreakpoint="1264"
  nexticon="$next"
  previcon="$prev"
>
  <v-skeleton-loader-stub
    class="my-1 mr-2"
    type="chip"
    types="[object Object]"
  />
  <v-skeleton-loader-stub
    class="my-1 mr-2"
    type="chip"
    types="[object Object]"
  />
  <v-skeleton-loader-stub
    class="my-1 mr-2"
    type="chip"
    types="[object Object]"
  />
  <v-skeleton-loader-stub
    class="my-1 mr-2"
    type="chip"
    types="[object Object]"
  />
  <v-skeleton-loader-stub
    class="my-1 mr-2"
    type="chip"
    types="[object Object]"
  />
  <v-skeleton-loader-stub
    class="my-1 mr-2"
    type="chip"
    types="[object Object]"
  />
  <v-skeleton-loader-stub
    class="my-1 mr-2"
    type="chip"
    types="[object Object]"
  />
  <v-skeleton-loader-stub
    class="my-1 mr-2"
    type="chip"
    types="[object Object]"
  />
  <v-skeleton-loader-stub
    class="my-1 mr-2"
    type="chip"
    types="[object Object]"
  />
  <v-skeleton-loader-stub
    class="my-1 mr-2"
    type="chip"
    types="[object Object]"
  />
</v-chip-group-stub>
`;
